contracts/src/mixins/SlashingMixin.sol

```diff
@@ -0,0 +1,154 @@
+// // SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.29;
+
+import {
+    PackedUserOperation
+} from "@account-abstraction/contracts/interfaces/PackedUserOperation.sol";
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IQuotaManager} from "../interfaces/QuotaManager.sol";
+import {PaymasterLib} from "../libs/PaymasterLib.sol";
+import {CoreLib} from "../libs/CoreLib.sol";
+import {IERC4337InfoProvider} from "../interfaces/ERC4337.sol";
+import {ISlashingCondition} from "../interfaces/Conditions.sol";
+import {InvalidUserOpPaymaster} from "../interfaces/Errors.sol";
+
+struct SlashingMixinConfig {
+    uint48 enableCoditionsDelay;
+    ISlashingCondition[] initialConditions;
+}
+
+abstract contract SlashingMixin is IQuotaManager, IERC4337InfoProvider {
+    using SafeERC20 for IERC20;
+    using PaymasterLib for PackedUserOperation;
+
+    struct ConditionStatus {
+        uint48 enabledAfter;
+        uint48 enabledUntil;
+    }
+
+    uint48 public immutable ENABLE_CONDITIONS_DELAY;
+
+    event Slashed(
+        ISlashingCondition indexed offendingCondition,
+        address indexed validator,
+        uint256 indexed amount
+    );
+
+    event EnableCondition(
+        ISlashingCondition indexed offendingCondition,
+        uint48 enabledFrom
+    );
+
+    event DisableCondition(
+        ISlashingCondition indexed offendingCondition,
+        uint48 enabledUntil
+    );
+
+    uint256 public slashedTokens;
+    mapping(ISlashingCondition => ConditionStatus) public enabledConditions;
+    mapping(bytes32 => uint256) public slashedUserOps;
+
+    constructor(SlashingMixinConfig memory _config) {
+        ENABLE_CONDITIONS_DELAY = _config.enableCoditionsDelay;
+        for (uint256 i = 0; i < _config.initialConditions.length; i++) {
+            enabledConditions[_config.initialConditions[i]] = ConditionStatus({
+                enabledAfter: uint48(block.timestamp),
+                enabledUntil: 0
+            });
+        }
+    }
+
+    function slashValidator(
+        ISlashingCondition offendingCondition,
+        PackedUserOperation calldata userOp
+    ) public {
+        // Only slashing for userOps related to this paymaster is allowed
+        require(
+            userOp.extractPaymaster() == address(this),
+            InvalidUserOpPaymaster()
+        );
+        (uint256 validAfter, uint256 validUntil) = userOp
+            .extractValidatorData();
+        ConditionStatus memory conditionStatus = enabledConditions[
+            offendingCondition
+        ];
+        require(
+            (conditionStatus.enabledAfter < validUntil || validUntil == 0) &&
+                (validAfter < conditionStatus.enabledUntil ||
+                    conditionStatus.enabledUntil == 0),
+            "Condition not active"
```

Comment:
nit: why mix errors with revert messages? I would prefer just using `error` types everywhere (they emit smaller code and are much easier to work with in general).

Reply:
Part of https://github.com/safe-research/harbour/issues/29